### ISSUE TITLE
[Firefox] How can I use a local file in Tampermonkey?

### ISSUE THREAD

**Issue by shu8** [Tampermonkey/tampermonkey#347]
I can't manage to `@require` a local file in Firefox.

For example, if I use this script:

```
// ==UserScript==
// @name         testing
// @match        *://*.stackoverflow.com/*
// @require      file:///C:/Users/shu8/Desktop/test.js
// @grant        GM_setValue
// ==/UserScript==
```

with `test.js` being only `alert('test');`, nothing happens.

However, the exact same script works in Chrome.

Is there a setting I need to change somewhere to allow local files, like I did in Chrome?

**Comment by derjanb** [comment]
Unfortunately this is not implemented by Firefox. Please vote for this issue or leave a comment there: https://bugzilla.mozilla.org/show_bug.cgi?id=1266960

Thanks.

**Comment by shu8** [comment]
@derjanb thanks for the quick reply! that's really annoying, I've voted on the bug, but it seems like I'm the only one!? Have I voted on the right place? :/

**Comment by derjanb** [comment]
@shu8 It's the only place I've found so far. You can also comment on this issue and explain why it is important for you.

**Comment by pmorch** [comment]
On the other hand, opening a file of "blob" in an external editor *is* supported by Firefox, and that is the main reason I'd like to use `@require file://` to begin with - as a workaround for editing in an external editor.

See e.g. [It's All Text!](https://addons.mozilla.org/en-US/firefox/addon/its-all-text/) that opens an external editor.

Perhaps in the meantime, the documentation could be updated, e.g. in [Make script editable in an external editor](https://forum.tampermonkey.net/viewtopic.php?p=1832#p1832) referenced by the [FAQ](http://tampermonkey.net/faq.php#Q402) and on the [Tampermonkey • Firefox](http://tampermonkey.net/?ext=dhdg&browser=firefox) page which in fact specifically mentions this as possible:
> Tampermonkey's editor is nice, but you do have your own, that is faster, better, ... The solution is to enable the file access and then you can @require the local copy of your script.

**Comment by svnpenn** [comment]
https://github.com/Tampermonkey/tampermonkey/issues/351

**Comment by AlfonsoML** [comment]
Besides Firefox, it seems that Edge is also affected, but I didn't find any existing issue.  
Does it already exist or am I wrong and it should work?

**Comment by pmorch** [comment]
Update: With the release of Firefox 57 (Quantum) this week, only WebExtensions are supported from now on.

The WebExtensions API does not allow access to local files, and probably never will. I believe this issue can therefore be closed, as "use a local file" is no longer possible in any extension.

In FF57, e.g. [It's All Text!](https://addons.mozilla.org/en-US/firefox/addon/its-all-text/) also doesn't work any more.

**Comment by tophf** [comment]
Quoting https://stackoverflow.com/a/44516256

>I finally found the way to do this using the Fetch requests and FileReader APIs.

We're using this method in Stylus for live-reload of usercss styles.

**Comment by derjanb** [comment]
@topfh Thanks for the tip. TM beta 4.5.4619 uses it. We'll see how long it is working. 

@all Please note that you have to enable script file access at Tampermonkey's options page first.

**Comment by svnpenn** [comment]
https://github.com/FirefoxBar/xStyle/issues/45

**Comment by svnpenn** [comment]
you all might want to look at this

https://github.com/FirefoxBar/xStyle/issues/45#issuecomment-346829585

**Comment by mgsloan** [comment]
How about having a small local file server and auto-fetching from it?  The server would just need to have liberal CORS settings, and ideally be configured to only serve your tamper monkey scripts. 

It really adds to the pain of hacking on userscripts if I have to manually copy them out of my text editor all the time.  Using a text editor inside a browser is just not ok.  I tried using the "edit-in-emacs" extension, but I could barely get it working even for github comments, never got it to work with tampermonkey's editor.

**Comment by trustable-code** [comment]
I tried it with a local web server, but was not able to include a script at runtime.
Only solution I know it to stay with Firefox 56 or to use Waterfox or Basilisk.
I'm happy with Waterfox.

**Comment by Hidden50** [comment]
A local web server worked for me on the initial load, though the script (included via ``@require``) was cached and then didn't update when I made changes.

I'm assuming the caching is done by tampermonkey. Maybe a non-caching version of ``@require`` might be possible?

**Comment by mgsloan** [comment]
FWIW, this is no longer necessary for me, because it turns out to be quite easy to convert a tampermonkey script into a browser extension.  The browser extensions then support easy reload from local files.  Here's the commit where I converted my project to a browser extension: https://github.com/mgsloan/todoist-shortcuts/commit/6875896ff7a802f287bc37284131eac8f4a5171d

**Comment by pmorch** [comment]
@Hidden50 : About caching: I think that standard browser caching is taking place. If the webserver doesn't send cache headers, it does the standard thing. I can't find a reference to it at the moment, but I seem to believe that "the standard caching thing" is to cache things for some `constant * age_of_resource`. (and my local webserver - `twistd -n web -p 8000 --path .` - does send the `Last-Modified` header from which `age_of_resource` can be derived).

So yeah, it is annoying and not perfect. What I've done is `touch` the files that I know will change, have a separate browser tab where that resource is open, clear the cache, then browser refresh. In the network tab of the debugger you can see whether the resource is cached or not. I don't think I've seen a situation where the networking tab disagreed with Firefox caching (I'm using Firefox). I'd be surprised if Chrome was any different.

Ideal would be to have a local webserver where you can control caching headers, so the browser doesn't cache at all, but I didn't bother because this worked well enough.

**Comment by derjanb** [comment]
Requiring local files is still working fine at Firefox up to version 60.0. You just have to enable the corresponding config option. (This is needed to not allow scripts local file access without a user's permission.)

![image](https://user-images.githubusercontent.com/767504/39910990-f5941f64-54f9-11e8-8d56-0d479d0400c7.png)

Note: requiring a local file may delay script execution a little bit. If this causes problems you can require a file from a local webserver. For development you can set "Externals" -> "Update Interval" to "Always". (Please use this in development only!) In order to allow fast script injection externals are updated everytime _after_ the script was executed.

![image](https://user-images.githubusercontent.com/767504/39911051-3af99412-54fa-11e8-9a0d-a4451edf3dcf.png)

**Comment by Hidden50** [comment]
Thanks. I completely missed the "update interval" setting. That solves the problem for me, I can work with this.

Local files still don't load in Firefox, even though I use the settings you describe (Firefox Quantum 60.0, ``Allow scripts to access local files: Externals (@require and @resource)``).
![image](https://user-images.githubusercontent.com/827830/39912326-ab7813d6-54fe-11e8-90cc-587fdcbd54c6.png)
The same settings and ``@require`` path work in chrome. I assume firefox denies access, though the way the message is phrased would also fit with a typo'd or otherwise invalid file path.

**Comment by derjanb** [comment]
Indeed, it's not working at Windows 10 anymore, but still working at Ubuntu. 🙄

**Comment by derjanb** [comment]
FYI: I've created another possible solution for script editing which is based on Tampermonkey's recently added WebDAV support. [Tampermonkey/tamperdav](https://github.com/Tampermonkey/tamperdav#readme)

**Comment by Ruko2010** [comment]
@derjanb is there any update on this issue? I use FF 62.0.3 and local file access is not possible...

**Comment by derjanb** [comment]
There is no news on this. FF 62.0.3 still allows local file access at Ubuntu. Other OS needs to wait for [Firefox bug 1266960](https://bugzilla.mozilla.org/show_bug.cgi?id=1266960).

**Comment by Ruko2010** [comment]
Thanks for the Update. I'll watch this Issue for progress.

**Comment by johnd0e** [comment]
@derjanb 
>  For development you can set "Externals" -> "Update Interval" to "Always". (Please use this in development only!) In order to allow fast script injection externals are updated everytime **after** the script was executed.

It's not convenient for some use cases.
May I ask you about alternative solutions?

- Add one more choice there: 'Always (before script)'
- Or 'Always (request in background)'
- Or add some kind of remote control to TM, to allow some external (node.js?) script to send command when it should update.

**Comment by derjanb** [comment]
> Or add some kind of remote control to TM, to allow some external (node.js?) script to send command when it should update.

@johnd0e Please take a look at [TamperDAV](https://github.com/Tampermonkey/tamperdav). It allows a script to be synced from disk every time it is modified.

**Comment by johnd0e** [comment]
Well, my userscripts are built from some kind source templates.
My current workflow is like this:
- change sources
- build userscripts
- goto Tampermonkey dashboard, press 'Update' button (and wait it to finish)
- reload the page

Is TamperDAV able to help me?
(I've tried it, and it has it's job, but not for my workflow)

**Comment by startsts** [comment]
July 2020,  still not implemented? Not working in latest Elementary OS

**Comment by knoajp** [comment]
Okay...

I managed to configure my workflow on Firefox/Tampermonkey with Keyboard Maestro; an automation app of macOS.
https://www.keyboardmaestro.com/main/

First, prepare a bookmarklet on Firefox for popupping an editor window with a dedicated title.
```javascript
(function(){
  const SUFFIX = ' [Local Script Updater]';
  const WIDTH = 320, HEIGHT = 640;
  const getTextarea = (d) => d.querySelector('[vis="true"] .CodeMirror-editor textarea');
  const getSaveButton = (d) => [...d.querySelectorAll('[vis="true"] ul.editormenu td.shortcut')].find(td => td.textContent === 'Cmd-S');
  const w = window.open(location.href, '_blank', `width=${WIDTH},height=${HEIGHT},resizable`);
  setTimeout(() => {
    const d = w.document, t = getTextarea(d), b = getSaveButton(d);
    d.title += SUFFIX;
    if(t && b){
      t.focus();
      t.addEventListener('change', e => b.click());
    }
  }, 5000);
})();
```

Second, prepare a macro on Keyboard Maestro for copying and pasting to the popped editor window.
I bound [F5] key with the macro for Sublime Text.
![20201010-231141](https://user-images.githubusercontent.com/844187/95657639-77caca00-0b50-11eb-94b1-17baa5a9a8dc.png)

How to use:
1. Open a script's editor page of Tampermonkey.
2. Pop a new editor window with the bookmarklet and leave it open.
3. Edit a script with an external editor.
4. press [F5] key to execute them all.

**Comment by xdhmoore** [comment]
My solution has been using a local nginx server. All my tampermonkey scripts have only the header metadata, like:

```
// ==UserScript==
// @name         myscriptname
// @namespace    http://tampermonkey.net/
// @version      1.007
// @description  try to take over the world!
// @author       You
// @match        https://www.example.com/*
// @require      http://localhost/userscripts/myscriptname/main.js
// @grant        none
// ==/UserScript==
```
Then I have to manually bump the version every time I update the script to break the cache. And I think sometimes reload the "example.com" page question twice. It's not ideal. I made a scaffolding tool that helps set it up and a deploy script to copy it from my dev folder into nginx, but the deploy script can't update the version above, so I have to bump it manually. Maybe there's a way to host the above metadata from my nginx instance as well. Then my deploy script could auto-increment the version and I could update everything in one go with my deploy script.

Anyway, there's probably a better way.

**Comment by mgsloan** [comment]
xdhmoore@ Nice approach!  You should be able to [disable caching in nginx](https://docs.nginx.com/nginx/admin-guide/content-cache/content-caching/#limiting-or-disabling-caching).  Alternatively, you could have the userscript [generate a script tag with a random query parameter](https://stackoverflow.com/a/7413275/1164871).

**Comment by xdhmoore** [comment]
I don't think I've tried fiddling with nginx cache headers. I'll have to try that. I don't think the random query parameter would help because it would need to go in the `// @require` statement of the metadata which is stored in Tampermonkey and AFAIK, not modifiable by script. Otherwise I would just bump the `// @version` there.

**Comment by TSSans-art** [comment]
Is it possible to @resource a non-javascript (.js) file from a file:// url?

**Comment by sausager** [comment]
I provide a good idea. this way don't need change any TamperMonkey setting.

You can create anew script like this
```
// ==UserScript==
// @name         New Userscript
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       You
// @match        http://www.website.com/*
// @grant        none
// ==/UserScript==

function addScript(url){
    var script = document.createElement('script');
    script.setAttribute('type','text/javascript');
    script.setAttribute('src',url);
    document.getElementsByTagName('head')[0].appendChild(script);
}
addScript('http://localhost:6868/index.js?t='+(new Date()).getTime())
```

Evey refresh, the page will be load javascript from http://localhost:6868/index.js

Create a simple loacal http server with nodejs or nginx etc, catch the url, output your code

Here is js code to create a simple http server with nodejs:

```
// http.js

const http = require('http')
const readFileSync = require('fs')['readFileSync']

http.createServer((req, res) => {
  if (req.url.startsWith('/index.js')) {
    const content = readFileSync('./index.js', 'utf8')
    res.writeHead(200, {
      'Content-Type': 'text/javascript; charset=utf-8'
    })
    res.end(content)
  } else {
    res.writeHead(404, {
      'Content-Type': 'text/plain; charset=utf-8'
    })
    res.end('error access')
  }
}).listen(6868)
```
Run command
`node http.js`

Now you can code `index.js` with any editer, enjoy~

**Comment by xdhmoore** [comment]
@sausager I was excited when I modified your example for my own use. However, I hit a problem today with a site that has a content security policy defined. While Tampermonkey I think has some plumbing to help with that for direct scripts, when I load my own script like this it gets blocked by the cross-origin CSP.

**Comment by NabiKAZ** [comment]
After 6 years in 2022... and still nothing! 👴

**Comment by derjanb** [comment]
> Unfortunately this is not implemented by Firefox. Please vote for this issue or leave a comment there: https://bugzilla.mozilla.org/show_bug.cgi?id=1266960
> 
> Thanks.

@NabiKAZ I assume you mean that it hasn't been fixed in Firefox yet. 😔

**Comment by NabiKAZ** [comment]
I wrote this code for my work:

```javascript
// ==UserScript==
// @resource     accounts_chrome file:///C:/Nabi/accounts.js
// @resource     accounts_firefox http://127.0.0.1:12345/accounts.js?111
// @grant        GM_getResourceText
// ==/UserScript==

//NOTE:
//  Chrome > Extensions > TamperMonkey > Allow access to file URLs
//  Firefox > php -S 0.0.0.0:12345

(function() {
    'use strict';

    var accounts = [];
    var accounts_chrome = GM_getResourceText("accounts_chrome");
    var accounts_firefox = GM_getResourceText("accounts_firefox");
    if (accounts_chrome) {
        accounts = JSON.parse(accounts_chrome);
    } else if (accounts_firefox) {
        accounts = JSON.parse(accounts_firefox);
    }

    alert(accounts);

})();
```
The content of `accounts.js` is json.
The `111` end of `accounts_firefox`  path, should be change for prevent caching.
For `Chrome`, Enable this: `Extensions > TamperMonkey > Allow access to file URLs`
For `Firefox`, In the `C:\Nabi\` path, run this: `php -S 0.0.0.0:12345`

**Comment by xdhmoore** [comment]
@NabiKAZ there's another ticket for the caching issue: https://github.com/Tampermonkey/tampermonkey/issues/723

**Comment by ZiuChen** [comment]
https://github.com/Tampermonkey/tampermonkey/issues/347#issuecomment-706614606

This method worked for my case.

**Comment by MotherOfClamperl** [comment]
#347 mostly works, but on pages like instagram it gives "Content Security Policy" errors.
eg: 
Loading failed for the <script> with source “http://localhost/userscripts/config.user.js?t=1660712589420”.

Content Security Policy: The page’s settings blocked the loading of a resource at http://localhost/userscripts/config.user.js?t=1660712589420 (“script-src”).[ line 14 > eval:19:45](https://www.instagram.com/%20line%2014%20%3E%20eval)

Content Security Policy: The page’s settings observed the loading of a resource at http://localhost/userscripts/config.user.js?t=1660712589420 (“script-src”). A CSP report is being sent.

Also, it hardly ever injects the script on the first page load. I keep having to refresh pages to get the scripts injected, especially when opening links in a new tab. I'm on macOS if it matters.

**Comment by rRobis** [comment]
I'm still experimenting with this new technique, and it seems that it's working (at least in my case). I make changes, refresh, and in firefox it gets changes on first reload 

I created a node.js server, that runs with nodemon like this:

```
cd "GRobe Scripts/GRobe/GRobeNode"
fnm use v19.9.0
nodemon app.js
```

In this app.js i have contents like this:

```
const express = require('express')
const cors = require('cors');
const fs = require('fs');
const path = require('path');
var process = require('process');

process.chdir('/home/input1a/GRobe Scripts/');

const util = require('util');
const readFile = util.promisify(fs.readFile);

const app = express()

const whitelist = ['YOURSITETHATYOUUSE.COM','127.0.0.1:3000'];

const corsOptions = {
  origin: function (origin, callback) {
    if (whitelist.indexOf(origin) !== -1) {
      callback(null, true)
    } else {
      callback(new Error(`${origin} Not allowed by CORS`))
    }
  }
}

// Enable CORS for the whitelist of sites
app.use(cors());
app.use(express.json())

// Make a endpoint for a script bundle
app.get('/GRobeGeneralBundle', async (req, res) => {
    const fileUrls = [
    'GRobe/Shared/smallFunctions.js',
    'GRobe/Shared/mousetrap.js',
    'GRobe/GRobeGeneral/GRobeGeneral.js',
    ];

    try {
        const combinedScript = await combineFilesSequentially(fileUrls);
        res.setHeader('Content-Type', 'application/javascript');
        res.send(combinedScript);
    } catch (error) {
        console.error('Error combining JavaScript files:', error);
        res.status(500).send('Internal Server Error');
    }
});

async function combineFilesSequentially(fileUrls) {
  let combinedScript = '';

  for (const fileUrl of fileUrls) {
    const fileData = await readFile(fileUrl, 'utf8');
    combinedScript += fileData;
  }

  return combinedScript;
}
app.listen(3000,()=>{
    console.log('Sever is running on localhost:3000')
})

```


And in the tampermonkey script i have it like this:

```
// ==UserScript==
// @name         GRobe™ General
// @namespace    GRobe
// @version      0.3
// @description  GRobe™
// @author       You
// @match        *://*/*
// @run-at       document-start
// @grant        GM_xmlhttpRequest
// ==/UserScript==

(()=>{const e="http://127.0.0.1:3000/GRobeGeneralBundle";const t=e=>{if(200===e.status&&e.responseText){eval(e.responseText)}else{console.error("Error loading Node.JS Script:",e.statusText)}};const n=()=>{GM_xmlhttpRequest({method:"GET",url:e,onload:t,onerror:t})};document.addEventListener("DOMContentLoaded",n)})();
```



The minified code is :
Basically it makes request to the endpoint and gets the combined script text and evaluates it, after the page has loaded
```
(function () {
	var nodeJSEndPoint = 'http://127.0.0.1:3000/GRobeGeneralBundle';

	// Function to evaluate the fetched script
	function evaluateScript(scriptContent) {
		eval(scriptContent);
	}

	// Function to handle the GM_xmlhttpRequest response
	function handleResponse(response) {
		if (response.status === 200 && response.responseText) {
			evaluateScript(response.responseText);
		} else {
			console.error('Error loading Node.JS Script:', response.statusText);
		}
	}

	// Function to make the GM_xmlhttpRequest
	function makeRequest() {
		GM_xmlhttpRequest({
			method: 'GET',
			url: nodeJSEndPoint,
			onload: handleResponse,
			onerror: handleResponse
		});
	}

	// Wait for the DOMContentLoaded event before making the request
	document.addEventListener('DOMContentLoaded', makeRequest);
})();
```

**Comment by derjanb** [comment]
You can now vote for https://bugzilla.mozilla.org/show_bug.cgi?id=1807608 as well. :-)

**Comment by Danielku15** [comment]
I just migrated to Brave and face the same problem that I cannot use local files. 

> `@require: couldn't load @require from URL 'file://C:\Users\Daniel\Dev\script.user.js': Access to this or all local files is forbidden!`

* Externals > Update Interval: Always
* Security > Allow scripts to access local files: Externals (@require and @resource)

Is it possible that Brave has disabled some setting compared to other Chromium based browsers? Any hints how I could get it working?

**Comment by simonkrauter** [comment]
``@require file:///`` works for me in Vivaldi (Chromium based too) with enabled "Allow access to file URLs".

**Comment by Danielku15** [comment]
@simonkrauter Thanks for the fast reply. I was missing the "Allow access to file URLs" in the browser extensions > Tampermonkey. It somehow didn't save, after some retries it was saved and now it works. 🥳

**Comment by kekkc** [comment]
Firemonkey implemented a solution for this in Firefox:
1. first userscript is created in Firemonkey and a huge 2MB data blob included (local variable) 
```
// @name PSL.user.js
```
3. userscript can then be deactivated
4. second userscript is created that needs to access PSL.user.js:
```
// @name 2ndScript
// @require PSL.user.js 
// @match file:///*/*
```

 it can access all the content/variables of the PSL userscript, even if its' deactivated

No performance impacting games via localstorage or GM_getValue/GM_setValue required ;)

**Comment by kekkc** [comment]
Workaround for Tampermonkey (seems Grant is not required):
```
// ==UserScript==
// @name		publicSuffixList.user.js
// @run-at		document-start
// @match		*://*/*
// @match		file:///*/*
// @allFrames	false
// ==/UserScript==

/* global unsafeWindow */
/* global GM_info */

const publicSuffixes = "data";
// Required for Tampermonkey
// eslint-disable-next-line camelcase
if (typeof GM_info !== 'undefined' && GM_info.scriptHandler === 'Tampermonkey') {
	unsafeWindow.publicSuffixes = publicSuffixes;
}
```

**Comment by kevinae102** [comment]
Firefox 153 added a dedicated, user-controlled “Access local files on your computer” permission for extensions in July 2026.

Does this new permission remove the Firefox limitation that originally blocked local file:/// resources in @require and @resource?

Tampermonkey 5.5.0 added File → Open for local userscripts in Firefox, but that appears different from the Chrome/Edge development workflow where a small DEV Loader continuously loads a VS Code source file using:

// @require file:///C:/path/to/script.js

Is Tampermonkey planning to use Firefox’s new local-file permission to support this workflow? If so, is there an issue or milestone tracking it?

**Comment by derjanb** [comment]
> Does this new permission remove the Firefox limitation that originally blocked local file:/// resources in @require and https://github.com/resource?

This new toggle only limits access to file:// URIs if not set. It is still not possible to access local files except they are opened in a tab. Programatically opening file: URIs is forbidden. Sorry.

Please vote for https://bugzilla.mozilla.org/show_bug.cgi?id=1807608 and https://bugzilla.mozilla.org/show_bug.cgi?id=1617594

> https://bugzilla.mozilla.org/show_bug.cgi?id=2034168
> We are currently not planning additional file access functionality ([bug 1246236](https://bugzilla.mozilla.org/show_bug.cgi?id=1246236), [bug 1807608](https://bugzilla.mozilla.org/show_bug.cgi?id=1807608)).
